### PR TITLE
Introduce IndexRoutingTable#readyForSearch

### PR DESCRIPTION
**Pull request by jakelandis**
This commit introduces IndexRoutingTable#readyForSearch to determine if a given index is available to for search.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java:299]: I removed this info level logger since it possible to run a system with only file based configuration and the security index is never created. In that case (which is the default state for serverless), this message could be emitted on every authc that needs to resolve role mappings.

**Comment by jakelandis** [comment]
labeling this as non-issue since it is only relevant for index/search shard splitting is not generally available

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: Does method require _every_ search shard copy to be active? It looks like `anyMatch` would return true even if only one of multiple shard copies of the same shard is active.

Additionally it might be good to avoid stream from within the loop

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:247]: This change could have potential issues. The SecurityIndexManager also manages token and profile indices besides the main security index. The profile index uses `fast_refresh` and does not have a separate search shard. I think this change would break most profile APIs, e.g. activation, suggestion. The token index can have separate search shards. But most of its operations are GET and do not rely on search shards. Do we want to block the GET access when the search shard is not available? Similarly, the main security index also sometimes uses the `isAvailable` check to guard GET access which can be blocked by the new check as well.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: > It looks like anyMatch would return true even if only one of multiple shard copies of the same shard is active.

That's the intention: this is to replace `allPrimaryShardsActive()` as a way to determine if an index is ready to search. But the name is confusing, I also interpret it as returning whether _all_ the search shard copies are active.

Can we call it something like `allShardsSearchReady` instead?

Also +1 to avoiding streams here, and also no need for `counter`, we can just `return false` as soon as we find any shard which isn't search-ready.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: ++ I would agree, this method is pretty suboptimal performance wise. It's fine if it's run like this for a limited set of indices but if it's executed with a wider scope we should either optimise it to avoid the stream usage or allocation through `org.elasticsearch.cluster.routing.IndexShardRoutingTable#shard` or move the code to where it's used for system indices only(right?).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (DaveCTurner) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:247]: I think it's kind of reasonable to consider the cluster as being in a fairly broken state if these indices do not have a search shard.

However I see that various places are checking `org.elasticsearch.xpack.security.support.SecurityIndexManager.State#indexHealth` to see whether to proceed, which is kind of the same as checking `allPrimaryShardsActive`, so they may need changing too.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: Also, I would like to note we should use `org.elasticsearch.cluster.routing.OperationRouting#canSearchShard()` which takes into account the new fast refresh setting in Serverless. And I would suggest to extend the test to have a fast refresh index. Maybe it's a good idea to also put a comment above the isSearchable() method that people should ideally use the canSearchShard() method, just to remind.

**Review by kingherc — COMMENTED** [review]
Consider fast refresh as well as commented.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: > But the name is confusing,

I see that now...sorry for the confusion. How about `readyForSearch` ? `all` is still a bit ambiguous w.r.t. every vs. needed

>  Additionally it might be good to avoid stream from within the loop / this method is pretty suboptimal performance wise / Also +1 to avoiding streams here

It is difficult to know that up front. The loop here will generally be small and it difficult to know how much of a performance hit using stream API actually is. Is it the stream allocation, it is the number of iteratation within the stream, how do we know that anyMatch from a stream is not more performant than manual iteration.  Additionally, the API only exposes this via a Stream.of (which is a pretty recent addition). I assume it is exposed as stream to avoid exposing the raw array. So to avoid the stream I would need to expose the raw array and to do that i would need to make an immutable copy ... but that seems as expensive as using a Stream. This code might be executed in a way that is extremely sensitive to performance (like frequent cluster state updates) but it is not today and don't think it's usage in that manner is likely (maybe it is?). I can probably find a way to avoid the stream or the copy but I question if that needed and/or if it is a premature optimization. 

> or move the code to where it's used for system indices only(right?).

I can move the code to only service my usecase(s) but thought this would be generally useful. 

> and also no need for counter, we can just return false as soon as we find any shard which isn't search-ready.

++

> Also, I would like to note we should use org.elasticsearch.cluster.routing.OperationRouting#canSearchShard() which takes into account the new fast refresh setting. 

Thanks for calling that out. I had not really considered fast refresh. I had assumed that since fast refresh indices don't have replicas the primary would be marked as both search and index which works fine (and is tested) with isSearchable().  Can you help me understand why that is not the case ?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: > how do we know that anyMatch from a stream is not more performant than manual iteration. 

lots of benchmarking for many-shards and such + the theory here is quite clear as well. Unless the whole thing inlines (which it won't here for standard JVM settings) you have far more code and virtual method calls to execute with the stream compared to a loop + have allocations that burn more cycles outside the main execution path.

> I can probably find a way to avoid the stream or the copy but I question if that needed and/or if it is a premature optimization.

Why would you need any copy here? You can just iterate like we do elsewhere via:

```java
    for (IndexShardRoutingTable shardRoutingTable : this.shards) {
            for (int idx = 0; copy < shardRoutingTable.size(); idx++) {
```

can't you?

>

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:247]: > The profile index uses fast_refresh and does not have a separate search shard.

I had not considered fast_refresh much since I assumed fast_refresh indices can not have replicas and thus will be always be marked as index and search shards. I asked for some clarification in the comment above and can switch over to use canSearchShard() instead of isSearchable(). 

> Do we want to block the GET access when the search shard is not available? 

Probably not, but it is an edge case. I don't think I know the exact behavior for a GET when a search shard is unavailable ... does it try the real time get, and but could still fail if the translog has already moved on ? how far back w.r.t. time/flush of the translog would the GET actually still work ? 

> to see whether to proceed, which is kind of the same as checking allPrimaryShardsActive, so they may need changing too.

Thanks for calling that one out.  I (or will ask someone) will revisit the security code base looking for other cases where the primary check is insufficient for a future PR.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java:240]: The copy comment was in reference to the array but I see now that we expose access to a single entry from the array. I will use that.

I believe and trust your assessment. The push back here is more about without your (and others in thread) assessment a) knowing when/where we should avoid streams b) the existence of APIs that expose the stream we are not supposed to use c) knowing the performance characteristics for alternatives to streams. 

I understand the motivation and need to keep things fast, but I am failing to know how to reason about this level of optimization in practice. When does the expected usage / area of code elevate itself to the point when I need critically think about how the code will be inlined and virtual method calls ? Do these optimization concerns extend beyond Stream usage ? Should there be warning in the code for these areas ? Short of running micro bench marks (or asking you :)) how do I know that the alternative is actually more performant and not a incorrect micro optimization?

**Comment by jakelandis** [comment]
PR is updated to no longer use a Stream and support fast refresh. I still have some meta questions w.r.t. optimization and fast refresh with replicas but have adopted the suggestions provided. I also did not make any changes to address the concern w.r.t. get workflows + unavailable search shards or additional checks that could be impacted by the index/search split. I will follow up on those (at least w.r.t. to the security workflows).

**Comment by jakelandis** [comment]
test failure was instance of https://github.com/elastic/elasticsearch/issues/99916

**Comment by jakelandis** [comment]
build failure [unrelated](https://github.com/elastic/elasticsearch/issues/99982).

@elasticsearchmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java]: > Can you help me understand why that is not the case ?

Unfortunately the `isSearchable()` function is depending on the shard's role which is an enum. While to know whether an index is a fast refresh index, we need to access the index metadata. So we created a new function canSearchShard that accesses both arguments to make the final judgement.

File: server/src/test/java/org/elasticsearch/cluster/routing/IndexRoutingTableTests.java
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/cluster/routing/IndexRoutingTableTests.java:120]: nit This is the correct behavior currently for fastRefresh; search/replica shards do not do anything. Indeed, it may change behavior in the future, but I believe this test should then fail when we try to do that, and we'll simply change it. Thus I am not sure the TODO is needed.

**Review by kingherc — APPROVED** [review]
LGTM. However, I would urge before merging:

- Running a serverless PR CI with this commit. Just to check that nothing broke beforehand.
- Getting at least one more approval, since I see there were several other co-reviewers and comments. Maybe from a security-related folk would be nice, since my knowledge on the SecurityIndexManager is limited.

**Comment by jakelandis** [comment]
thanks @kingherc ! 

> Running a serverless PR CI with this commit. Just to check that nothing broke beforehand.

will do

> Getting at least one more approval, since I see there were several other co-reviewers and comments. Maybe from a security-related folk would be nice, since my knowledge on the SecurityIndexManager is limited.

@ywangd - would you mind doing a full review ?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:246]: This might be a semantic change. The method is used to populated the `isAvailable` value. Today is available means all primary shards are functional in stateful. It means available for both read and write. The proposed change means the index can be searchable (read) but not necessary writable because primary shards might not be active. There is a test case for it in this PR
> 2 unassigned primaries that are index only with some replicas that are all available

I wonder whether we should instead tighten this check to be both available and searchable instead of just searchable since there are [use cases](https://github.com/elastic/elasticsearch/blob/554b64e36ccc4859254fd26700dd68f532ecb61e/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L1882-L1884) where we use this check for write purpose. Or maybe we should have separate methods for available and searchable checks. But that may be a larger change. 

Another minor issue is that when we [report error](https://github.com/elastic/elasticsearch/blob/567ab3a29996dc31d3c8039a74309346f72669d8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java#L138-L141) for `isAvailable == false`, it will still say "at least one primary shard is unavailable" even though it might not be the case.

**Review by ywangd — COMMENTED** [review]
I left a question for whether replacing isAvailable with isSearchable is a semantic change.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:246]: > The proposed change means the index can be searchable (read) but not necessary writable because primary shards might not be active.

In stateful: All replicas can be promoted to primary and are marked as DEFAULT so either the primary or replica will pass the `isSearchable()` check. The `isAvailable()` check is correct for writes except for the period of time after primary failure but before replica promotion. If there are no replicas to promote then the `isAvailable()` check will (correctly) return false. Since this check is more of an optimization / log helper it is does not need to be precisely correct during that short period.  Also, the setup for the test referenced is only applicable for stateless since it uses index/search only shards . 

In stateless: This is more relevant since as mentioned we use this check for both read and write and the shards are not all DEFAULT. However, read dominates the usage both by occurrence in the code and frequency of calls. The check is currently broken for reads which means that we are looking at primary shards to know if we can read. The consequence of that is reduced availability (for read requests when index shards are unavailable) and excessive logging that this PR fixes that. Fixing the read check comes at the cost of breaking the write check with the consequence is reduced availability (for write requests when search shards are unavailable ). Checking for both read/write would defeat the goal of this PR resulting in reduced availability for both read and write workflows. I think it is worth the effort split the check up into isAvailableForSearch, isAvailableForIndex, and isAvailableForRealTimeGet since in serverless each have their own semantics and can result in reduced availability if/when the index/search shards are unavailable. I was hoping to do that in a follow up PR since that is a large for this PR and this fixes the dominate issue. If you prefer I can pull the security changes here and handle it all once, but I do have a minor preference to proceed as-is with a follow up PR to handle the refactoring. 

>  it will still say "at least one primary shard is unavailable" even though it might not be the case.

Good catch ... will fix.

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:246]: > Checking for both read/write would defeat the goal of this PR resulting in reduced availability for both read and write workflows.

This prompted me to look into the context of this PR and I think it is ES-6842. IIUC, we have two excessive logs that this PR aims to fix:
1. (10%) "The security index is not yet available - no role mappings can be loaded "
2. (90%) "failed to load role mappings from index [.security] skipping all mappings."

For the first logging message, I think the fix is incorrect because the intention is that we don't want this log message when the security index is not "created". Therefore I think the fix should be adding a check for `indexExists()`. We call this method as a separate check in all other places (see this [example](https://github.com/elastic/elasticsearch/blob/ccc896d1284cb7a5fb7713110aedc1f8976f45c8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L1642-L1643)) but it is missed for `getMappings()`. If the security index has not been created at all, it is not an exceptional situation and we always just return an empty result set. Concretely I am suggesting the following change to the `getMappings` method as the fix
```diff
     private void getMappings(ActionListener<List<ExpressionRoleMapping>> listener) {
-        if (securityIndex.isAvailable()) {
+        if (securityIndex.indexExists() == false) {
+            listener.onResponse(List.of());
+        } else if (securityIndex.isAvailable()) {
             loadMappings(listener);
         } else {

```

For the second logging message, I took a closer look at how primary and replica shards behave in edge cases. My conclusion is that the new `readyForSearch()` check actually behaves as what I suggested [above](https://github.com/elastic/elasticsearch/pull/99890#discussion_r1341268755), i.e. it checks _both_ available and searchable. When a primary shard fails, if no replica can be promoted to the new primary, [all replicas will be moved to "Unassigned"](https://github.com/elastic/elasticsearch/blob/ccc896d1284cb7a5fb7713110aedc1f8976f45c8/server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java#L659-L664). An unassigned shard will fail to pass the new `readyForSearch()` check which requires a shard to be active in the first place. Put it in another way, we cannot expect GET or Search to keep working when the security index has only search-only shards. Related, an index-only shard is *not* able to serve GET requests (realtime or not) when there is no search shards because the request fails earlier on at routing phase due to no search shards (whether this is really what we want in long term is a separate question). In summary, for GET or Search to reliably work, _both_ index and search shards are required to be available. There is no availability optimisation we can _currently_ have to serve either request with just an index shard or just a search shard.

Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:246]: Thanks for digging in deep on this ... w.r.t. availability concerns I see you are correct, but still find it surprising that the availability of index_only shards impacts the availability of the search_only shards.  I suspect that is not what we want in the long term and will adjust (the security bits) of this PR accordingly.  I misinterpreted a comment that led me down this path and then applied some faulty assumptions. Thanks again for setting me straight ! (I'll take another look at the log message too.)

--- tangential ...
> the availability of index_only shards impacts the availability of the search_only shards

I don't think the inverse applies ... if the search_only shards are unavailable but the index_only shards can still be available. This (edge case) scenario could result in failure of read your writes via real time get where the write succeeds but the get fails. Not related here, but noteworthy.

**Comment by jakelandis** [comment]
I decided to remove remove the consuming code from this PR and integrate will integrate the security code to use readyForSearch in a followup PR. I will merge this once CI passes and put up the follow up PR in the next day or two.

**Comment by jakelandis** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2-fips

(another case of https://github.com/elastic/elasticsearch/issues/99943)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:246]: This change is still security related. Do you mean to keep it in for this PR?

**Review by ywangd — APPROVED** [review]
LGTM

Thanks for the iteration.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:246]: doh ! ... nope.. will revert that change too.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:245]:
```suggestion
        if (routingTable == null || routingTable.allPrimaryShardsActive() == false) {
```

**Comment by jakelandis** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1

(seem like an infrastructure issue)

**Pull request merged**
